### PR TITLE
Feature/warmstart

### PR DESCRIPTION
**Pull request by bkleyn**
New functionality to:
* Warm start untrained (cold) arms 
* Remove existing arms

Main changes include addition of `warm_start` and `remove_arm` methods to MAB, with corresponding changes to BaseMAB and the individual bandit implementations. To enable warm start, each bandit has to implement a `_copy_arms` method specifying what attributes should be copied from a warm arm to cold arm.

Documentation and tests have been updated.

### CODE REVIEW

File: mabwiser/neighbors.py
Comment (dorukkilitcioglu) [reply mabwiser/neighbors.py:77]: Also, I might've missed it, but is there a test for this behavior?

**Review by dorukkilitcioglu — DISMISSED** [review]
Looks good overall, thanks for making this a part of MABWiser! My only concrete suggestion is to explicitly note the difference between regular warm start and neighborhood warm start in docstring and the online documentation. I assume you'll be building and updating the docs after getting the approvals, G2G after that.

File: mabwiser/base_mab.py
Comment (bkleyn) [reply mabwiser/base_mab.py:294]: It uses `_get_arm_distances` which is a method of the class?

File: mabwiser/base_mab.py
Comment (bkleyn) [reply mabwiser/base_mab.py]: Good suggestion. [Updated](https://github.com/fidelity/mabwiser/pull/47/commits/a62f8cacb0c8178df5814790a03501f7594892ce).

File: tests/test_base.py
Comment (bkleyn) [reply tests/test_base.py]: Thanks, fixed.

File: mabwiser/mab.py
Comment (bkleyn) [reply mabwiser/mab.py:993]: I removed the "check arm is in" and the "check arm is not in" from `self._validate_arm` so that I can use it for both `add_arm` and `remove_arm`. Alternatively, I would have had to use separate validation methods for add and remove.

File: mabwiser/base_mab.py
Comment (dorukkilitcioglu) [reply mabwiser/base_mab.py:293]: That's also static, so you could say `BaseMAB._get_arm_distances` instead of `self._get_arm_distances`, right? Not sure if IDEs would complain that `BaseMAB` isn't initialized though.

Comment (bkleyn) [reply mabwiser/base_mab.py]: Updated.

File: mabwiser/base_mab.py
Comment (bkleyn) [reply mabwiser/base_mab.py:192]: Yes, that's correct.

File: mabwiser/neighbors.py
Comment (bkleyn) [reply mabwiser/neighbors.py:77]: Yes, that's correct. The nearest neighbor policy only fits the learning policy when a prediction is made for a new context. We can only determine which arms are untrained (cold) after the policy has been fit, at which point we can execute warm start.

I'll update documentation as suggested + add specific warm start test for this behavior.

Comment (bkleyn) [reply mabwiser/neighbors.py:77]: Pushed updates.

**Comment by bkleyn** [comment]
> Looks good overall, thanks for making this a part of MABWiser! My only concrete suggestion is to explicitly note the difference between regular warm start and neighborhood warm start in docstring and the online documentation. I assume you'll be building and updating the docs after getting the approvals, G2G after that.

Thanks for the review @dorukkilitcioglu. I've addressed all your questions and comments. I'll hold off on updating the website docs until @skadio has had a chance to review.

**Review by skadio — APPROVED** [review]

**Pull request merged**
